File: src/github.rs
//! Github API query of issues.

use Result;
use futures::{Future, Stream};
use hyper::{Client, Method, Request};
use hyper::header::{Authorization, Bearer, UserAgent};
use hyper_tls::HttpsConnector;
use serde_json;
use std::env;
use std::io;
use tokio_core::reactor::Core;

static API_URL: &'static str = "https://api.github.com/graphql";
static QUERY: &'static str = r#"
{
  search(query: "assignee:d3zd3z is:issue repo:zephyrproject-rtos/zephyr",
    type: ISSUE, last:100) {
    edges {
      node {
        ... on Issue {
          title
          url
          state
          number
        }
      }
    }
  }
}"#;

// The json request.
#[derive(Debug, Serialize)]
struct Query {
    query: String,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all="snake_case")]
enum QueryResult {
    Data(Search)
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all="snake_case")]
enum Search {
    Search(Edges),
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all="snake_case")]
enum Edges {
    Edges(Vec<Edge>),
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all="snake_case")]
enum Edge {
    Node(Node),
}

#[derive(Debug, Deserialize, Serialize, Clone)]
struct Node {
    title: String,
    state: String,
    url: String,
    number: u64,
}

impl QueryResult {
    /// Get the nodes out of the deep structure.
    fn into_nodes(self) -> Vec<Node> {
        match self {
            QueryResult::Data(search) => match search {
                Search::Search(edges) => match edges {
                    Edges::Edges(nodes) => {
                        nodes.into_iter().map(|node| match node {
                            Edge::Node(node) => node,
                        }).collect()
                    }
                }
            }
        }
    }
}

pub fn query_zephyr() -> Result<()> {
    let token = env::var("GITHUB_TOKEN").expect("Must set GITHUB_TOKEN in env");

    let mut core = Core::new()?;
    let handle = core.handle();
    let client = Client::configure()
        .connector(HttpsConnector::new(4, &handle)?)
        .build(&handle);

    let uri = API_URL.parse()?;

    let mut req = Request::new(Method::Post, uri);
    {
        let headers = req.headers_mut();
        headers.set(Authorization(Bearer {
            token: token,
        }));
        headers.set(UserAgent::new("d3zd3z-jirasync"));
    }

    let query = Query {
        query: QUERY.to_string(),
    };
    req.set_body(serde_json::to_string(&query)?);
    /*
    let mut query = serde_json::Map::new();
    query.insert("query".to_string(), Value::String(QUERY.to_string()));
    req.set_body(serde_json::to_string(&Value::Object(query))?);
    */

    let work = client.request(req).and_then(|res| {
        println!("Response: {}", res.status());

        res.body().concat2().and_then(move |body| {
            let v: QueryResult = serde_json::from_slice(&body).map_err(|e| {
                io::Error::new(
                    io::ErrorKind::Other,
                    e
                )
            })?;

            // Flatten the json.
            let nodes = v.into_nodes();

            dump_github(nodes);
            Ok(())
        })
    });
    core.run(work)?;

    Ok(())
}

fn dump_github(mut nodes: Vec<Node>) {
    // Sort by issue number.  TODO: Get milestone.
    nodes.sort_by_key(|a| a.number);

    println!("Issues at [https://github.com/zephyrproject-rtos/zephyr/]\n");
    println!("||Issue||Description||Status||");
    for node in &nodes {
        let color = match node.state.as_ref() {
            "OPEN" => "red",
            "CLOSED" => "green",
            _ => "blue",
        };
        println!("|[{}|{}]|{}|{{color:{}}}{}{{color}}|",
                 node.number,
                 node.url,
                 super::escape(&node.title),
                 color, node.state);
    }
}

File: cmd/jirasync/github.go
package main

import (
	"context"
	"fmt"
	"os"
	"sort"

	"github.com/shurcooL/githubql"
	"golang.org/x/oauth2"
)

// Github queries
type githubProject struct {
	user string
	repo string
	desc string
}

func (prj *githubProject) Desc() string {
	return prj.desc
}

// The information we want from a single issue.
type issue struct {
	Title  githubql.String
	Url    githubql.String
	State  githubql.String
	Number githubql.Int
}

func (prj *githubProject) Query() error {
	// Taken from githubql's README
	src := oauth2.StaticTokenSource(
		&oauth2.Token{AccessToken: os.Getenv("GITHUB_TOKEN")},
	)
	httpClient := oauth2.NewClient(context.Background(), src)

	client := githubql.NewClient(httpClient)

	var query struct {
		Search struct {
			Edges []struct {
				Node struct {
					Issue issue `graphql:"... on Issue"`
				}
			}
			PageInfo struct {
				EndCursor   githubql.String
				HasNextPage githubql.Boolean
			}
		} `graphql:"search(query:$query, type:ISSUE, first:100, after:$cursor)"`
	}

	queryText := fmt.Sprintf("assignee:%s is:issue repo:%s", prj.user, prj.repo)
	variables := map[string]interface{}{
		"query":  githubql.String(queryText),
		"cursor": (*githubql.String)(nil),
	}

	var issues []issue

	for {
		err := client.Query(context.Background(), &query, variables)
		if err != nil {
			return err
		}

		for _, edge := range query.Search.Edges {
			issues = append(issues, edge.Node.Issue)
		}
		// fmt.Printf("Page: %+v\n", query.Search.PageInfo)
		if !query.Search.PageInfo.HasNextPage {
			break
		}
		variables["cursor"] = githubql.NewString(query.Search.PageInfo.EndCursor)
	}

	// Sort by issue number for consistent results.
	sort.Sort(ByIssueNumber(issues))

	fmt.Printf("Issues at [https://github.com/%s/]\n\n", prj.repo)
	fmt.Printf("||Issue||Description||Status||\n")
	for i := range issues {
		var color string
		switch issues[i].State {
		case "OPEN":
			color = "red"
		case "CLOSED":
			color = "green"
		default:
			color = "blue"
		}
		fmt.Printf("|[%d|%s]|%s|{color:%s}%s{color}|\n",
			issues[i].Number,
			issues[i].Url,
			jiraEscape(string(issues[i].Title)),
			color,
			issues[i].State)
	}

	return nil
}
